openssl-tests/src/ffdhe_kx_with_openssl.rs

```diff
@@ -0,0 +1,239 @@
+use std::{
+    fs::{self, File},
+    io::{BufReader, Read, Write},
+    net::{TcpListener, TcpStream},
+    str,
+    sync::Arc,
+    thread,
+};
+
+use rustls::{
+    crypto::{ring::default_provider, CryptoProvider},
+    version::{TLS12, TLS13},
+    ClientConfig, RootCertStore, ServerConfig, SupportedProtocolVersion,
+};
+use rustls_pemfile::Item;
+use rustls_pki_types::{CertificateDer, PrivateKeyDer};
+
+use crate::{
+    ffdhe::{self, FfdheKxGroup},
+    utils::verify_openssl3_available,
+};
+
+use openssl::ssl::{SslAcceptor, SslFiletype, SslMethod};
+
+const CERT_CHAIN_FILE: &str = "../test-ca/rsa/end.fullchain";
+const PRIV_KEY_FILE: &str = "../test-ca/rsa/end.key";
+const CA_FILE: &str = "../test-ca/rsa/ca.der";
+const CA_PEM_FILE: &str = "../test-ca/rsa/ca.cert";
+
+fn root_ca() -> RootCertStore {
```

Comment:
I think these helpers should be below the `#[test]` instances that use them to match the [top-down ordering convention](https://github.com/rustls/rustls/blob/main/CONTRIBUTING.md#top-down-ordering-within-modules)